File: api-rails/app/controllers/users_controller.rb
# frozen_string_literal: true

class UsersController < ApplicationController
  def index
    @users = User.all
    render json: @users
  end

  def show
    @user = User.find params[:id]
    render @user
  end

  def create
    @user = User.new user_params
    if @user.save
      render json: @user, status: :ok
    else
      render json: @user.errors, status: :unprocessable_entity
    end
  end

  def update
    @user = User.find params[:id]
    if @user.update(user_params)
      render json: @user, status: :ok
    else
      render json: @user.errors, status: :unprocessable_entity
    end
  end

  def destroy
    @user = User.find params[:id]
    @user.destroy
    render json: { message: 'User has been destroyed' }, status: :ok
  end

  private

  def user_params
    params.require(:user).permit(:name, :age, :dob, :nationality)
  end
end

File: api_express/src/controllers/users_controller.ts
import { User } from "../types/models"

import * as UserRecord from "../models/user"
import { Request, Response } from "express"

const index = (async (req: Request, res: Response): Promise<void> => {
  const users: User[] = await UserRecord.all()
  res.json(users)
})

const show = (async (req: Request, res: Response): Promise<void> => {
  const id: number = Number(req.params.id)
  const user: User | null = await UserRecord.find(id)

  if (!user) {
    res.status(404).send("User not found")
  } else {
    res.json(user)
  }
})

const create = (async (req: Request, res: Response): Promise<void> => {
  const data: User = req.body
  const user: User = await UserRecord.create(data)
  res.json(user)
})

const update = (async (req: Request, res: Response): Promise<void> => {
  const id: number = Number(req.params.id)
  const user: User | null = await UserRecord.find(id)

  if(!user) {
    res.status(404).send("User not found")
  } else {
    const data: User = req.body
    const user: User = await UserRecord.update(id, data)
    res.json(user)
  }
})

const destroy = (async (req: Request, res: Response): Promise<void> => {
  const id: number = Number(req.params.id)
  const user: User | null = await UserRecord.find(id)

  if(!user) {
    res.status(404).send("User not found")
  } else {
    await UserRecord.destroy(id)
    res.send("User deleted successfully")
  }
})

export { index, show, create, update, destroy }
